source/common/http/http1/codec_impl.cc

```diff
@@ -785,7 +793,13 @@ Envoy::StatusOr<int> ConnectionImpl::onHeadersCompleteBase() {
 }
 
 void ConnectionImpl::bufferBody(const char* data, size_t length) {
-  buffered_body_.add(data, length);
+  auto slice = current_dispatching_buffer_->frontSlice();
+  if (data == slice.mem_ && length == slice.len_) {
+    buffered_body_.move(*current_dispatching_buffer_, length);
+    dispatching_slice_already_drained_ = true;
```

Comment:
I think could be a future refinement, but it is admittedly more complicated:

```
if (data + length == slice.mem_ + slice.len_) {
  if (data != slice.mem_) {
    buffered_body_.drain(data - slice.mem_);
  }
  buffered_body_.move(*current_dispatching_buffer_, length);
  dispatching_slice_already_drained_ = true;
} else {
  ...
}
```

Reply:
Thinking about this more, I think your current heuristic is better than the one I'm suggesting here.  For the one I suggested to work we need something like lenght > 2048 or 4096 as an additional criteria for the move.